File: src/main/python/WNLMainWindow.py
from PyQt5.QtWidgets import (QMainWindow, QVBoxLayout, QHBoxLayout, QPushButton, QWidget, QGridLayout, QLabel, QListWidget, QListWidgetItem)
from PyQt5.QtGui import QIcon
from PyQt5.QtCore import Qt
from WNLAboutDialog import WNLAboutDialog
from WNLSoundGrabber import WNLSoundGrabber
from WNLPlaybackManager import WNLPlaybackManager

# Class for main application window
class WNLMainWindow(QMainWindow):
    def __init__(self, appctxt, *args, **kwargs):
        super().__init__(*args, **kwargs)

        # Set the application context variable
        self.appctxt = appctxt

        # Set window and application title
        self.setWindowTitle("White Noise Linux")

        # Set default main window size
        self.resize(800, 600)

        # Set main window minimum size
        self.setMinimumSize(800, 600)

        # Create the central widget for the window
        self.centralWidget = QWidget()

        # Create main layout widget
        self.mainLayout = QVBoxLayout()
        self.mainLayout.setAlignment(Qt.AlignTop)

        # Create top bar layout
        self.topBarLayout = QHBoxLayout()
        self.topBarLayout.setAlignment(Qt.AlignRight)

        # Populate top bar with buttons
        self.populateTopBar()

        # Add the top bar to the main layout
        self.mainLayout.addLayout(self.topBarLayout)

        # Create sound selection layout
        self.soundSelectionLayout = QGridLayout()
        self.soundSelectionLayout.setAlignment(Qt.AlignHCenter)

        # Populate sound selection layout
        self.populateSoundSelect()

        # Add sound selection to main layout
        self.mainLayout.addLayout(self.soundSelectionLayout)

        # Create media controls layout
        self.controlsLayout = QHBoxLayout()
        self.controlsLayout.setAlignment(Qt.AlignCenter)
        self.controlsLayout.setContentsMargins(0, 10, 0, 10)

        # Populate media controls layout
        self.populateControls()

        # Add media controls layout to main layout
        self.mainLayout.addLayout(self.controlsLayout)

        # Set the main layout as the layout for the window
        self.centralWidget.setLayout(self.mainLayout)
        self.setCentralWidget(self.centralWidget)

    def populateTopBar(self):
        # Add the "About" button
        about = QPushButton()
        about.setFixedSize(25, 25)
        about.setIcon(QIcon.fromTheme("help-about"))
        about.setToolTip("About")
        about.clicked.connect(self.openAboutDialog)
        self.topBarLayout.addWidget(about)

    def openAboutDialog(self):
        aboutDialog = WNLAboutDialog(self.appctxt)
        aboutDialog.exec_()

    def populateSoundSelect(self):
        # Create and populate the QVBoxLayout for selecting available sounds
        availSoundsLayout = QVBoxLayout()

        availSoundsLabel = QLabel("Available Sounds:")
        availSoundsSelect = QListWidget()

        availSoundsLayout.addWidget(availSoundsLabel)
        availSoundsLayout.addWidget(availSoundsSelect)

        self.getAvailableSounds()

        # Create and populate the QVBoxLayout for moving sounds to the currently playing box
        moveSoundsWidget = QWidget()
        moveSoundsWidget.setFixedHeight(100)

        moveSoundsLayout = QVBoxLayout()

        addSound = QPushButton()
        addSound.setFixedSize(75, 25)
        addSound.setIcon(QIcon.fromTheme("go-next"))
        addSound.setToolTip("Add Sound to Currently Playing")
        addSound.setEnabled(False)
        addSound.clicked.connect(self.addSound)

        rmSound = QPushButton()
        rmSound.setFixedSize(75, 25)
        rmSound.setIcon(QIcon.fromTheme("go-previous"))
        rmSound.setToolTip("Remove Sound from Currently Playing")
        rmSound.setEnabled(False)
        rmSound.clicked.connect(self.rmSound)

        moveSoundsLayout.addWidget(addSound)
        moveSoundsLayout.addWidget(rmSound)

        moveSoundsWidget.setLayout(moveSoundsLayout)

        # Create and populate the QVBoxLayout for displaying currently playing sounds
        currSoundsLayout = QVBoxLayout()

        currSoundsLabel = QLabel("Currently Playing:")
        currSoundsSelect = QListWidget()

        currSoundsLayout.addWidget(currSoundsLabel)
        currSoundsLayout.addWidget(currSoundsSelect)

        # Add all the layouts to the sound selection layout
        self.soundSelectionLayout.addLayout(availSoundsLayout, 0, 0, 1, 1)
        self.soundSelectionLayout.addWidget(moveSoundsWidget, 0, 1, 1, 1)
        self.soundSelectionLayout.addLayout(currSoundsLayout, 0, 2, 1, 1)

        # Set the slot for when an item in the available sounds list is selected, so that the add button is enabled
        self.soundSelectionLayout.itemAt(0).itemAt(1).widget().selectionModel().selectionChanged.connect(self.handleAvailListChange)

        # Set the slot for when an item in the currently playing sounds list is selected, so that the remove button is enabled
        self.soundSelectionLayout.itemAt(2).itemAt(1).widget().selectionModel().selectionChanged.connect(self.handlePlayingListChange)

        # Populate the list of sounds
        self.populateAvailableSounds()

    def getAvailableSounds(self):
        soundGrabber = WNLSoundGrabber()
        
        # Get the sounds from file
        self.sounds = soundGrabber.getSounds()

        # Instantiate the playback manager
        self.playbackManager = WNLPlaybackManager(self.sounds)

    def populateAvailableSounds(self):
        # Add sounds to the available sounds list
        for sound in self.sounds:
            newItem = QListWidgetItem()
            newItem.setText(sound.name)
            newItem.setData(Qt.UserRole, sound)
            self.soundSelectionLayout.itemAt(0).itemAt(1).widget().addItem(newItem)

    def addSound(self):
        # Get the sound item and remove it from the available sounds list
        soundItem = self.soundSelectionLayout.itemAt(0).itemAt(1).widget().takeItem(
            self.soundSelectionLayout.itemAt(0).itemAt(1).widget().selectionModel().selectedIndexes()[0].row()
        )

        # Add the sound to the list of currently playing sounds
        self.soundSelectionLayout.itemAt(2).itemAt(1).widget().addItem(soundItem)

        # If the currently playing list now has 8 sounds (the maximum), disable the "available" list
        if self.soundSelectionLayout.itemAt(2).itemAt(1).widget().count() >= 8:
            self.soundSelectionLayout.itemAt(0).itemAt(1).widget().selectionModel().clear()
            self.soundSelectionLayout.itemAt(0).itemAt(1).widget().setEnabled(False)

        # Tell the playback manager to add the sound to the list
        self.playbackManager.addSound(soundItem.data(Qt.UserRole))

    def rmSound(self):
        # Get the sound item and remove it from the currently playing sounds list
        soundItem = self.soundSelectionLayout.itemAt(2).itemAt(1).widget().takeItem(
            self.soundSelectionLayout.itemAt(2).itemAt(1).widget().selectionModel().selectedIndexes()[0].row()
        )

        # Add the sound to the list of available sounds
        self.soundSelectionLayout.itemAt(0).itemAt(1).widget().addItem(soundItem)

        # If the currently playing list now has less than 8 sounds (the maximum), enable the "available" list
        if self.soundSelectionLayout.itemAt(2).itemAt(1).widget().count() < 8:
            self.soundSelectionLayout.itemAt(0).itemAt(1).widget().setEnabled(True)

        # Tell the playback manager to remove the sound from the list
        self.playbackManager.rmSound(soundItem.data(Qt.UserRole))

    def handleAvailListChange(self, currentSelection):
        # If the number of selected sounds is greater than 0, enable the add button, otherwise disable it
        if len(currentSelection.indexes()) > 0:
            self.soundSelectionLayout.itemAt(1).widget().layout().itemAt(0).widget().setEnabled(True)
        else:
            self.soundSelectionLayout.itemAt(1).widget().layout().itemAt(0).widget().setEnabled(False)

    def handlePlayingListChange(self, currentSelection):
        # If the number of selected sounds is greater than 0, enable the remove button, otherwise disable it
        if len(currentSelection.indexes()) > 0:
            self.soundSelectionLayout.itemAt(1).widget().layout().itemAt(1).widget().setEnabled(True)
        else:
            self.soundSelectionLayout.itemAt(1).widget().layout().itemAt(1).widget().setEnabled(False)

    def populateControls(self):
        # Create the play/pause button
        playPause = QPushButton()
        playPause.setFixedSize(50, 50)
        playPause.setIcon(QIcon.fromTheme("media-playback-start"))
        playPause.setToolTip("Start Playing Audio")
        playPause.clicked.connect(self.handlePlayPauseClick)

        # Add the play/pause button to the media controls layout
        self.controlsLayout.addWidget(playPause)

    def handlePlayPauseClick(self):
        # Change the icon from play to pause, or vice versa, depending on the current status, and change playback accordingly
        if self.controlsLayout.itemAt(0).widget().icon().name() == "media-playback-start":
            # Change the icon
            self.controlsLayout.itemAt(0).widget().setIcon(QIcon.fromTheme("media-playback-pause"))
            self.controlsLayout.itemAt(0).widget().setToolTip("Pause Audio")

            # Play audio
            self.playbackManager.playAudio()
        else:
            # Change the icon
            self.controlsLayout.itemAt(0).widget().setIcon(QIcon.fromTheme("media-playback-start"))
            self.controlsLayout.itemAt(0).widget().setToolTip("Start Playing Audio")

            # Pause the audio
            self.playbackManager.pauseAudio()

File: src/wnlmainwindow.cpp
// Copyright 2021 Ryan Simmons
// This program is distributed under the terms of the GNU General Public License Version 3

#include "wnlmainwindow.h"
#include "ui_wnlmainwindow.h"

WNLMainWindow::WNLMainWindow(QWidget *parent)
    : QMainWindow(parent),
      ui(new Ui::WNLMainWindow),
      dBusConn(QDBusConnection::sessionBus())
{
    ui->setupUi(this);

    // Get the sounds from disk
    WNLSoundGrabber soundGrabber;
    QFuture<QVector<WNLSound>> soundsFuture = QtConcurrent::run(soundGrabber, &WNLSoundGrabber::getSounds);

    QFutureWatcher<QVector<WNLSound>>* soundsFutureWatcher = new QFutureWatcher<QVector<WNLSound>>;
    connect(soundsFutureWatcher, &QFutureWatcher<QVector<WNLSound>>::finished, [=]()
    {
        // Once the available sounds have been read from disk, add them to the available sounds widget
        for (WNLSound sound : soundsFutureWatcher->result())
        {
            QListWidgetItem* soundItem = new QListWidgetItem(sound.getTitle());
            soundItem->setData(Qt::UserRole, QVariant::fromValue<WNLSound>(sound));
            ui->soundSelectionWidget->findChild<QListWidget *>("availSoundsSelect", Qt::FindChildrenRecursively)->addItem(soundItem);
        }

        // Setup D-Bus stuff
        this->setupDBus();

        delete soundsFutureWatcher;
    });
    soundsFutureWatcher->setFuture(soundsFuture);

    // Setup the playlists dropdown
    QtConcurrent::run(this, &WNLMainWindow::setupPlaylists);
}

WNLMainWindow::~WNLMainWindow()
{
    delete ui;

    // Unregister the MPRIS object
    this->dBusConn.unregisterObject("/org/mpris/MediaPlayer2");

    // Unregister the MPRIS service
    bool unregisteredSuccessfully = this->dBusConn.unregisterService(QStringLiteral("org.mpris.MediaPlayer2.whitenoiselinux.instance%1").arg(QCoreApplication::applicationPid()));
    if (unregisteredSuccessfully == false)
    {
        qDebug() << "Unable to unregister MPRIS service.";
    }
}

void WNLMainWindow::playAudio()
{
    // Change the icon
    ui->mediaControlsWidget->findChild<QPushButton *>("playPauseButton", Qt::FindChildrenRecursively)->setIcon(QIcon::fromTheme("media-playback-pause"));

    // Play audio
    this->playbackManager.playAudio();

    // Tell MPRIS that audio has started playing
    this->playerAdaptor->setPlaybackStatus("Playing");
}

void WNLMainWindow::pauseAudio()
{
    // Change the icon
    ui->mediaControlsWidget->findChild<QPushButton *>("playPauseButton", Qt::FindChildrenRecursively)->setIcon(QIcon::fromTheme("media-playback-start"));

    // Pause audio
    this->playbackManager.pauseAudio();

    // Tell MPRIS that the audio has been paused
    this->playerAdaptor->setPlaybackStatus("Paused");
}

void WNLMainWindow::on_aboutButton_clicked() const
{
    // Create new about dialog that deletes itself on close
    WNLAboutDialog *aboutDialog = new WNLAboutDialog();
    aboutDialog->setAttribute(Qt::WA_DeleteOnClose);

    // Position the about dialog in the centre of the parent window
    aboutDialog->move(this->geometry().center() - aboutDialog->rect().center());

    // Show the dialog (modal)
    aboutDialog->exec();
}

void WNLMainWindow::on_addSoundButton_clicked()
{
    // Get pointers to both lists
    QListWidget* availSoundsSelect = ui->soundSelectionWidget->findChild<QListWidget *>("availSoundsSelect", Qt::FindChildrenRecursively);
    QListWidget* currSoundsSelect = ui->soundSelectionWidget->findChild<QListWidget *>("currSoundsSelect", Qt::FindChildrenRecursively);

    // Move all selected sounds to the "currently playing" list
    for (QModelIndex selectedIndex : availSoundsSelect->selectionModel()->selectedIndexes())
    {
        this->addSound(selectedIndex.row());
    }

    // Update the MPRIS track title value
    this->playerAdaptor->setTrackTitle(this->playbackManager.getCurrentlyPlayingString());
}

void WNLMainWindow::addSound(int row)
{
    // Get pointers to both lists
    QListWidget* availSoundsSelect = ui->soundSelectionWidget->findChild<QListWidget *>("availSoundsSelect", Qt::FindChildrenRecursively);
    QListWidget* currSoundsSelect = ui->soundSelectionWidget->findChild<QListWidget *>("currSoundsSelect", Qt::FindChildrenRecursively);

    // Remove item from the available sounds list
    QListWidgetItem* selectedItem = availSoundsSelect->takeItem(row);

    // Add the item to the currently playing list
    currSoundsSelect->addItem(selectedItem);

    // Open the sound and add it to the list of currently playing sounds
    this->playbackManager.addSound(selectedItem->data(Qt::UserRole).value<WNLSound>());
}

void WNLMainWindow::on_rmSoundButton_clicked()
{
    // Get pointers to both lists
    QListWidget* currSoundsSelect = ui->soundSelectionWidget->findChild<QListWidget *>("currSoundsSelect", Qt::FindChildrenRecursively);
    QListWidget* availSoundsSelect = ui->soundSelectionWidget->findChild<QListWidget *>("availSoundsSelect", Qt::FindChildrenRecursively);

    // Move all selected sounds to the "available sounds" list
    for (QModelIndex selectedIndex : currSoundsSelect->selectionModel()->selectedIndexes())
    {
        this->rmSound(selectedIndex.row());
    }

    // Update the MPRIS track title value
    this->playerAdaptor->setTrackTitle(this->playbackManager.getCurrentlyPlayingString());
}

void WNLMainWindow::rmSound(int row)
{
    // Get pointers to both lists
    QListWidget* currSoundsSelect = ui->soundSelectionWidget->findChild<QListWidget *>("currSoundsSelect", Qt::FindChildrenRecursively);
    QListWidget* availSoundsSelect = ui->soundSelectionWidget->findChild<QListWidget *>("availSoundsSelect", Qt::FindChildrenRecursively);

    // Remove item from the currently playing list
    QListWidgetItem* selectedItem = currSoundsSelect->takeItem(row);

    // Add the item to the available sounds list
    availSoundsSelect->addItem(selectedItem);

    // Remove the sound from the playback manager
    this->playbackManager.rmSound(row);
}

void WNLMainWindow::on_currSoundsSelect_itemSelectionChanged()
{
    // If the number of items selected in the list is greater than 0, enable the button for moving sounds back to the available list
    if (ui->soundSelectionWidget->findChild<QListWidget *>("currSoundsSelect", Qt::FindChildrenRecursively)->selectedItems().count() > 0)
    {
        ui->soundSelectionWidget->findChild<QPushButton *>("rmSoundButton", Qt::FindChildrenRecursively)->setEnabled(true);
    }
    else // Otherwise, disable it
    {
        ui->soundSelectionWidget->findChild<QPushButton *>("rmSoundButton", Qt::FindChildrenRecursively)->setEnabled(false);
    }
}

void WNLMainWindow::on_availSoundsSelect_itemSelectionChanged()
{
    // If the number of items selected in the list is greater than 0, enable the button for moving sounds to the currently playing list
    if (ui->soundSelectionWidget->findChild<QListWidget *>("availSoundsSelect", Qt::FindChildrenRecursively)->selectedItems().count() > 0)
    {
        ui->soundSelectionWidget->findChild<QPushButton *>("addSoundButton", Qt::FindChildrenRecursively)->setEnabled(true);
    }
    else // Otherwise, disable it
    {
        ui->soundSelectionWidget->findChild<QPushButton *>("addSoundButton", Qt::FindChildrenRecursively)->setEnabled(false);
    }
}

void WNLMainWindow::on_playPauseButton_clicked()
{
    // Change the icon from play to pause, or vice versa, depending on the current status, and change playback accordingly
    if (this->playbackManager.isPaused())
    {
        // Call the playAudio method
        this->playAudio();
    }
    else
    {
        // Call the pauseAudio method
        this->pauseAudio();
    }
}

void WNLMainWindow::setupDBus()
{
    // Instantiate both adaptors (multiple are needed to implement multiple interfaces)
    // The first adaptor's pointer doesn't need to be saved to a variable since we won't need to access it again
    // Qt will handle the memory management
    new WNLMPRISMediaPlayer2Adaptor(this);
    this->playerAdaptor = new WNLMPRISMediaPlayer2PlayerAdaptor(this);

    //Connect the player adaptor's signals to slots in the application so that the application can react when various events occur
    connect(this->playerAdaptor, &WNLMPRISMediaPlayer2PlayerAdaptor::wasPaused, this, &WNLMainWindow::pauseAudio);
    connect(this->playerAdaptor, &WNLMPRISMediaPlayer2PlayerAdaptor::wasPlayPaused, this, &WNLMainWindow::on_playPauseButton_clicked);
    connect(this->playerAdaptor, &WNLMPRISMediaPlayer2PlayerAdaptor::wasPlayed, this, &WNLMainWindow::playAudio);

    // Register the MPRIS service
    bool registeredSuccessfully = this->dBusConn.registerService(QStringLiteral("org.mpris.MediaPlayer2.whitenoiselinux.instance%1").arg(QCoreApplication::applicationPid()));
    if (registeredSuccessfully == false)
    {
        qDebug() << "Unable to register MPRIS service, media controls from Linux won't work.";
    }

    // Register the MPRIS object
    registeredSuccessfully = this->dBusConn.registerObject("/org/mpris/MediaPlayer2", this);
    if (registeredSuccessfully == false)
    {
        qDebug() << "Unable to register MPRIS object, media controls from Linux won't work.";
    }
}

void WNLMainWindow::on_saveButton_clicked()
{
    // Get a pointer to the currently playing sounds list
    QListWidget* currSoundsSelect = ui->soundSelectionWidget->findChild<QListWidget *>("currSoundsSelect", Qt::FindChildrenRecursively);

    // Get a pointer to the playlist combo box
    QComboBox* playlistComboBox = ui->topBarWidget->findChild<QComboBox *>("playlistComboBox", Qt::FindChildrenRecursively);

    // Make a new QVector to hold all the currently playing sounds
    QVector<WNLSound> playingSounds;

    // Add all the currently playing sounds to the vector
    for (int i = 0; i < currSoundsSelect->count(); i++)
    {
        QListWidgetItem* listItem = currSoundsSelect->item(i);
        WNLSound sound = listItem->data(Qt::UserRole).value<WNLSound>();
        playingSounds.append(sound);
    }

    // Save the currently playing sounds as a playlist
    this->playlistManager.savePlaylist(playlistComboBox->currentText(), playingSounds);

    // Clear focus
    this->ui->centralwidget->setFocus();
}

void WNLMainWindow::on_deleteButton_clicked()
{
    // Get a pointer for the playlist combo box
    QComboBox* playlistComboBox = ui->topBarWidget->findChild<QComboBox *>("playlistComboBox", Qt::FindChildrenRecursively);

    // Get the WNLPlaylist object for the currently selected playlist
    WNLPlaylist deletedPlaylist = playlistComboBox->currentData(Qt::UserRole).value<WNLPlaylist>();

    // Display a message box, where the user confirms whether or not they want to delete the playlist
    QMessageBox::StandardButton response;
    response = QMessageBox::question(this, "Delete Playlist", "Are you sure you wish to delete playlist \"" + deletedPlaylist.getName() + "\"?", QMessageBox::Yes | QMessageBox::No);

    // If the user clicked yes, delete the playlist
    if (response == QMessageBox::Yes)
    {
        // Remove the playlist from the playlist manager and delete it from disk
        if (this->playlistManager.deletePlaylist(deletedPlaylist))
        {
            // Remove the playlist from the playlist combo box
            playlistComboBox->removeItem(playlistComboBox->currentIndex());
        }
    }

    // Clear focus
    this->ui->centralwidget->setFocus();
}

void WNLMainWindow::addPlaylist(WNLPlaylist newPlaylist)
{
    // Get a pointer for the playlist combo box
    QComboBox* playlistComboBox = ui->topBarWidget->findChild<QComboBox *>("playlistComboBox", Qt::FindChildrenRecursively);

    // Add the new playlist to the combo box
    playlistComboBox->addItem(newPlaylist.getName(), QVariant::fromValue<WNLPlaylist>(newPlaylist));
}

void WNLMainWindow::setupPlaylists()
{
    // Setup directory structure
    this->playlistManager.setupDirStructure();

    // Grab playlists
    this->playlistManager.grabPlaylists();

    // Add each playlist to the playlists combo box
    QComboBox* playlistComboBox = ui->topBarWidget->findChild<QComboBox *>("playlistComboBox", Qt::FindChildrenRecursively);

    // Add each playlist to the combo box
    for (WNLPlaylist playlist : this->playlistManager.getPlaylists())
    {
        playlistComboBox->addItem(playlist.getName(), QVariant::fromValue<WNLPlaylist>(playlist));
    }

    // Connect the playlist added signal to the add playlist method
    connect(&(this->playlistManager), &WNLPlaylistManager::playlistAdded, this, &WNLMainWindow::addPlaylist);
}

void WNLMainWindow::on_loadButton_clicked()
{
    // Get a pointer for the playlist combo box
    QComboBox* playlistComboBox = ui->topBarWidget->findChild<QComboBox *>("playlistComboBox", Qt::FindChildrenRecursively);

    // Asynchronously run the method to load a playlist
    QtConcurrent::run(this, &WNLMainWindow::loadPlaylist, playlistComboBox->currentData().value<WNLPlaylist>());

    // Clear focus
    this->ui->centralwidget->setFocus();
}

void WNLMainWindow::loadPlaylist(WNLPlaylist playlist)
{
    // Get pointers to both lists
    QListWidget* currSoundsSelect = ui->soundSelectionWidget->findChild<QListWidget *>("currSoundsSelect", Qt::FindChildrenRecursively);
    QListWidget* availSoundsSelect = ui->soundSelectionWidget->findChild<QListWidget *>("availSoundsSelect", Qt::FindChildrenRecursively);

    // Get a pointer for the playlist combo box
    QComboBox* playlistComboBox = ui->topBarWidget->findChild<QComboBox *>("playlistComboBox", Qt::FindChildrenRecursively);

    int currCount = currSoundsSelect->count();

    // Remove all sounds currently playing
    for (int i = 0; i < currCount; i++)
    {
        this->rmSound(0);
    }

    // Add all the available items in the list widget to a list (for some reason, Qt doesn't provide a function to get this)
    QVector<QListWidgetItem*> availableListItems;
    for (int i = 0; i < availSoundsSelect->count(); i++)
    {
        availableListItems.append(availSoundsSelect->item(i));
    }

    // Add all of the sounds in the selected playlist to the currently playing list
    for (WNLSound sound : playlistComboBox->currentData(Qt::UserRole).value<WNLPlaylist>().getSounds())
    {
        // Filter the list of items to get the item that corresponds to the current sound from the playlist
        // Since there's the possibility that two different sound files have the same title property, we need to compare their file paths as well, instead of just their titles
        // This unfortunately complicates the logic in this section since we can't just run availSoundsSelect->findItems()
        QVector<QListWidgetItem*> resultingSoundList = QVector<QListWidgetItem*>::fromList(QtConcurrent::blockingFiltered(availableListItems, [=](const QListWidgetItem* item)
        {
            return item->data(Qt::UserRole).value<WNLSound>() == sound;
        }).toList());

        // If the sound was not found, print an error message, otherwise add it to the currently playing sounds
        if (resultingSoundList.size() == 0)
        {
            qDebug() << "Unable to find file " << sound.getFilePath() << ".";
        }
        else
        {
            this->addSound(availSoundsSelect->row(resultingSoundList.at(0)));
        }
    }
}
